Can not run fuzzer

## Bug Report

<!-- Thanks for your bug report! Don't worry if you can't fill out all the sections. -->

### What version of TiKV are you using?
<!-- You can run `tikv-server --version` -->
master

### What operating system and CPU are you using?
<!-- If you're using Linux, you can run `cat /proc/cpuinfo` -->
x86_x64

### Steps to reproduce
<!-- If possible, provide a recipe for reproducing the error. A complete runnable program is good. -->
1. run `cargo run --package fuzz -- run Afl fuzz_codec_bytes`
2. see
```
12T08:58:49.336Z] error[E0599]: no method named `generate_files` found for type `&Builder` in the current scope
[2020-02-12T08:58:49.336Z]   --> /rust/git/checkouts/protobuf-build-174b08b051947a10/42e52b9/src/lib.rs:56:14
[2020-02-12T08:58:49.336Z]    |
[2020-02-12T08:58:49.336Z] 56 |         self.generate_files();
[2020-02-12T08:58:49.336Z]    |              ^^^^^^^^^^^^^^ method not found in `&Builder`
[2020-02-12T08:58:49.336Z] 
[2020-02-12T08:58:49.336Z] error: aborting due to previous error
[2020-02-12T08:58:49.336Z] 
[2020-02-12T08:58:49.336Z] For more information about this error, try `rustc --explain E0599`.
[2020-02-12T08:58:49.336Z] error: could not compile `protobuf-build`.
[2020-02-12T08:58:49.336Z] warning: build failed, waiting for other jobs to finish...
[2020-02-12T08:59:16.192Z] error: build failed
[2020-02-12T08:59:16.192Z] Running fuzzer failed: error building afl instrumented binary, exit code Some(101)
[2020-02-12T08:59:16.192Z] Error: 1
```

### What did you expect?
No error and start fuzz

### What did happened?

--- a/Makefile
+++ b/Makefile
@@ -226,7 +226,9 @@ test:
 	export DYLD_LIBRARY_PATH="${DYLD_LIBRARY_PATH}:${LOCAL_DIR}/lib" && \
 	export LOG_LEVEL=DEBUG && \
 	export RUST_BACKTRACE=1 && \
-	cargo test --no-default-features --features "${ENABLE_FEATURES}" --all --exclude tests --exclude cdc ${EXTRA_CARGO_ARGS} -- --nocapture && \
+	cargo test --no-default-features --features "${ENABLE_FEATURES}" --all --exclude tests --exclude \
+		cdc --exclude fuzz-targets --exclude fuzzer-honggfuzz --exclude fuzzer-afl --exclude fuzzer-libfuzzer \
+		${EXTRA_CARGO_ARGS} -- --nocapture && \
 	cargo test --no-default-features --features "${ENABLE_FEATURES}" -p tests --bench misc ${EXTRA_CARGO_ARGS} -- --nocapture && \
 	cd components/cdc && cargo test --no-default-features --features "${ENABLE_FEATURES}" -p cdc ${EXTRA_CARGO_ARGS} -- --nocapture && cd ../.. && \
 	if [[ "`uname`" == "Linux" ]]; then \
@@ -244,12 +246,15 @@ test:
 
 # This is used for CI test
 ci_test: ci_doc_test
-	cargo test --no-default-features --features "${ENABLE_FEATURES}" --all --exclude tests --exclude cdc --exclude cdc --all-targets --no-run --message-format=json
+	cargo test --no-default-features --features "${ENABLE_FEATURES}" --all --exclude tests --exclude cdc \
+		--exclude fuzz-targets --exclude fuzzer-honggfuzz --exclude fuzzer-afl --exclude fuzzer-libfuzzer \
+		--all-targets --no-run --message-format=json
 	cd tests && cargo test --no-default-features --features "${ENABLE_FEATURES}" --no-run --message-format=json
 	cd components/cdc && cargo test --no-default-features --features "${ENABLE_FEATURES}" --no-run --message-format=json
 
 ci_doc_test:
-	cargo test --no-default-features --features "${ENABLE_FEATURES}" --all --exclude tests --doc
+	cargo test --no-default-features --features "${ENABLE_FEATURES}" --all --exclude tests --exclude cdc \
+		--exclude fuzz-targets --exclude fuzzer-honggfuzz --exclude fuzzer-afl --exclude fuzzer-libfuzzer --doc
 
 ## Static analysis
 ## ---------------
@@ -280,6 +285,7 @@ ALLOWED_CLIPPY_LINTS=-A clippy::module_inception -A clippy::needless_pass_by_val
 # PROST feature works differently in test cdc and backup package, they need to be checked under their folders.
 clippy: pre-clippy
 	@cargo clippy --all --exclude cdc --exclude backup \
+		--exclude fuzz-targets --exclude fuzzer-honggfuzz --exclude fuzzer-afl --exclude fuzzer-libfuzzer \
 		--all-targets --no-default-features \
 		--features "${ENABLE_FEATURES}" -- $(ALLOWED_CLIPPY_LINTS)
 	@for pkg in "cdc" "backup"; do \
@@ -287,6 +293,11 @@ clippy: pre-clippy
 		cargo clippy -p $$pkg --all-targets --no-default-features \
 			--features "${ENABLE_FEATURES}" -- $(ALLOWED_CLIPPY_LINTS) && \
 		cd ../.. ;\
+	done
+	@for pkg in "fuzz" "fuzz/fuzzer-afl" "fuzz/fuzzer-honggfuzz" "fuzz/fuzzer-libfuzzer"; do \
+		cd $$pkg && \
+		cargo clippy --all-targets -- $(ALLOWED_CLIPPY_LINTS) && \
+		cd -; \
 	done
 
 # TODO fix tests warnings
@@ -306,6 +317,12 @@ pre-audit:
 audit: pre-audit
 	cargo audit
 
+FUZZER ?= Honggfuzz
+
+.PHONY: fuzz
+fuzz:
+	@cargo run --package fuzz --no-default-features --features "${ENABLE_FEATURES}" -- run ${FUZZER} ${FUZZ_TARGET} \
+	|| echo "" && echo "Set the target for fuzzing using FUZZ_TARGET and the fuzzer using FUZZER (default is Honggfuzz)"
 
 ## Special targets
 ## ---------------

--- a/fuzz/cli.rs
+++ b/fuzz/cli.rs
@@ -250,13 +250,13 @@ fn run_afl(target: &str) -> Result<(), Error> {
 fn run_honggfuzz(target: &str) -> Result<(), Error> {
     pre_check(
         Command::new("cargo").args(&["hfuzz", "version"]),
-        "cargo install hfuzz --version 0.5.34",
+        "cargo install honggfuzz --version 0.5.45",
     )?;
 
     let fuzzer = Fuzzer::Honggfuzz;
 
     let mut rust_flags = env::var("RUSTFLAGS").unwrap_or_default();
-    rust_flags.push_str("-Z sanitizer=address");
+    rust_flags.push_str(" -Z sanitizer=address");
 
     let hfuzz_args = format!(
         "-f {} \

--- a/fuzz/targets/Cargo.toml
+++ b/fuzz/targets/Cargo.toml
@@ -4,14 +4,18 @@ version = "0.0.1"
 publish = false
 edition = "2018"
 
-[features]
-prost-codec = ["tikv_util/prost-codec"]
-
 [lib]
 path = "mod.rs"
 
+# Due to some complexity in the way fuzz targets are built, we need to depend on some upstream crates
+# here in order to pass the "protobuf-codec" feature. If you want to fuzz with Prost, you'll need to
+# s/protobuf-codec/prost-codec/g
 [dependencies]
 byteorder = "1"
 failure = "0.1"
+grpcio = { version = "0.5.0-alpha.5", features = [ "openssl-vendored", "protobuf-codec" ], default-features = false }
+kvproto = { git = "https://github.com/pingcap/kvproto.git", default-features = false, features = [ "protobuf-codec" ] }
+protobuf-build = { version = "0.10", default-features = false, features = ["grpcio-protobuf-codec"] }
 tidb_query = { path = "../../components/tidb_query" }
 tikv_util = { path = "../../components/tikv_util" }
+tipb = { git = "https://github.com/pingcap/tipb.git", default-features = false, features = [ "protobuf-codec" ] }

